crates/polars-core/src/series/arithmetic/horizontal.rs

```diff
@@ -0,0 +1,223 @@
+    let non_null_cols = columns
+        .iter()
+        .filter(|x| x.dtype() != &DataType::Null)
+        .map(|c| c.as_materialized_series())
+        .collect::<Vec<_>>();
+
+    match non_null_cols.len() {
+        0 => {
+            if columns.is_empty() {
+                Ok(None)
+            } else {
+                // all columns are null dtype, so result is null dtype
+                Ok(Some(columns[0].clone()))
+            }
+        },
+        1 => Ok(Some(
+            apply_null_strategy(
+                if non_null_cols[0].dtype() == &DataType::Boolean {
+                    non_null_cols[0].cast(&DataType::UInt32)?
+                } else {
+                    non_null_cols[0].clone()
+                },
+                null_strategy,
+            )?
+            .into(),
+        )),
+        2 => sum_fn(
+            non_null_cols[0].clone(),
+            non_null_cols[1].clone(),
+            null_strategy,
+        )
+        .map(Column::from)
+        .map(Some),
+        _ => {
+            // the try_reduce_with is a bit slower in parallelism,
+            // but I don't think it matters here as we parallelize over columns, not over elements
+            let out = POOL.install(|| {
+                non_null_cols
+                    .into_par_iter()
+                    .cloned()
+                    .map(Ok)
+                    .try_reduce_with(|l, r| sum_fn(l, r, null_strategy))
+                    // We can unwrap because we started with at least 3 columns, so we always get a Some
+                    .unwrap()
+            });
+            out.map(Column::from).map(Some)
+        },
+    }
+}
+
+/// Compute the mean of all values horizontally across columns.
+///
+/// All columns need to be the same length or a scalar.
+pub fn mean_horizontal(
+    columns: &[Column],
+    null_strategy: NullStrategy,
+) -> PolarsResult<Option<Column>> {
+    let (numeric_columns, non_numeric_columns): (Vec<_>, Vec<_>) = columns.iter().partition(|s| {
+        let dtype = s.dtype();
+        dtype.is_numeric() || dtype.is_decimal() || dtype.is_bool() || dtype.is_null()
+    });
+
+    if !non_numeric_columns.is_empty() {
+        let col = non_numeric_columns.first().cloned();
+        polars_bail!(
+            InvalidOperation: "'horizontal_mean' expects numeric expressions, found {:?} (dtype={})",
+            col.unwrap().name(),
+            col.unwrap().dtype(),
+        );
+    }
+    let columns = numeric_columns.into_iter().cloned().collect::<Vec<_>>();
+    match columns.len() {
+        0 => Ok(None),
+        1 => Ok(Some(match columns[0].dtype() {
+            dt if dt != &DataType::Float32 && !dt.is_decimal() => {
+                columns[0].cast(&DataType::Float64)?
+            },
+            _ => columns[0].clone(),
+        })),
+        _ => {
+            let sum = || sum_horizontal(&columns, null_strategy);
+            let null_count = || {
+                columns
+                    .par_iter()
```

Comment:
Do we need this in core, if we could store it in `polars-ops` that wuold be preferred.

Reply:
It would remove the `sum_horizontal`, `mean_horizontal`, ... methods from the DataFrame methods. If you are okay with that, I can do it.

Reply:
We could `TraitExt` them in ops?